### PR TITLE
Resolve #1: Implement Hash Table with Chaining

### LINKED ISSUE #1: Implement HashTable with Chaining
Methods:
- insert
- remove
- search

Test cases needed

### PR DESCRIPTION
**Pull request by dvquy13**
Closes #1 

- Add tests
- Initiate gitignore
- TODO: convert arbitrage key to number

### CODE REVIEW

**Comment by dvquy13** [comment]
@trungducc It seems like without write permission to the base repo, I cannot link this PR to the issue #1 . Could you help me with the linking?

**Comment by dvquy13** [comment]
@toet251 Please review. Also, I need help implementing the function to convert any key to a number for hashing purpose. @RT1918 hey Tuan Anh would you like to help?

File: src/hash_table.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+from __future__ import annotations
+from typing import Union
+import random
+
+
+class Node:
```
Comment (trungducc) [inline src/hash_table.py]: Ideally, `Node` should be a private class inside `ChainHashTable`.

File: src/hash_table.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        self.key = key
+        self.val = val
+        self.next = next
+
+
+class ChainHashTable:
+    def __init__(self, n: int):
+        self.__bucket = [Node(None, None)] * n
+
+    def insert(self, key: str, val: Union[str, float]) -> bool:
```
Comment (trungducc) [inline src/hash_table.py]: This method will always return `True`. I think we don't need to return any value from this method.

File: src/hash_table.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        node: Node = self.__bucket[self._hash(key)]
+        while node.next:
+            node = node.next
+            if node.key == key:
+                node.val = val
+                return True
+        node.next = Node(key, val)
+        return True
+
+    def remove(self, key: str) -> bool:
```
Comment (trungducc) [inline src/hash_table.py]: Same as above, we don't need to return `bool` from this method.

File: src/hash_table.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        node: Node = self.__bucket[self._hash(key)]
+        while node and node.next:
+            next_node = node.next
+            if next_node.key == key:
+                node.next = next_node.next
+                return True
+            node = next_node
+        return False
+
+    def search(self, key: str) -> float:
```
Comment (trungducc) [inline src/hash_table.py]: As I understand, value can be `string`. Is it ok if we only return `float`?

File: src/hash_table.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        node: Node = self.__bucket[self._hash(key)]
+        while node:
+            if node.key == key:
+                return node.val
+            node = node.next
+        return -1
```
Comment (trungducc) [inline src/hash_table.py]: `None` should be returned if key doesn't exists

File: tests/test_hash_table.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import unittest
+
+from src.hash_table import ChainHashTable
+
+
+class TestChainHashTable(unittest.TestCase):
```
Comment (trungducc) [inline tests/test_hash_table.py:6]: We should also check for edge cases. For example, search for key which doesn't exist or remove from empty hash table

**Review by trungducc — CHANGES_REQUESTED** [review]
Please add comments for methods purpose.

**Comment by trungducc** [comment]
Look like we are missing Table Doubling. It should be implemented as a part of this PR.

File: src/hash_table.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+class Node:
+    def __init__(self, key: str, val: Union[str, float], next: Node = None):
+        self.key = key
+        self.val = val
+        self.next = next
+
+
+class ChainHashTable:
+    def __init__(self, n: int):
+        self.__bucket = [Node(None, None)] * n
```
Comment (toet251) [inline src/hash_table.py]: Dont use this to generate list with type object. This will use the object `reference` for all elements of list. So when u change self.__bucket[0] then all the other self.__bucket[i] will be affected.
It's OK for [0] * size, but not [object type] * size

File: src/hash_table.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            node = node.next
+            if node.key == key:
+                node.val = val
+                return True
+        node.next = Node(key, val)
+        return True
+
+    def remove(self, key: str) -> bool:
+        node: Node = self.__bucket[self._hash(key)]
+        while node and node.next:
```
Comment (toet251) [inline src/hash_table.py:29]: I dont think we need to check node.next
This while never reach to the last element, so the last element never be removed.

File: src/hash_table.py [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+    def insert(self, key: str, val: Union[str, float]) -> bool:
+        node: Node = self.__bucket[self._hash(key)]
+        while node.next:
+            node = node.next
+            if node.key == key:
+                node.val = val
+                return True
+        node.next = Node(key, val)
+        return True
```
Comment (toet251) [inline src/hash_table.py:25]: I think this function is still working. But i dont think it's the best for while node.next, if you dont process node.next in while logic, then it seems u always dont care about the last element, which have `next` is None.

File: src/hash_table.py [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        key: int = self._convert_key_to_num(key)
+        return key % len(self.__bucket)
+
+    def _convert_key_to_num(self, key: str) -> int:
+        # TODO: Update this function, for now it assumes all key can be
+        # convert to number
+        try:
+            return int(key)
+        except:
+            return round(random.random() * 100)
```
Comment (toet251) [inline src/hash_table.py]: I use Python built-in function `hash` for this ^^ u can try it

Comment (toet251) [reply src/hash_table.py]: https://docs.python.org/3/library/functions.html#hash

File: src/hash_table.py (hunk 1 above)
Comment (dvquy13) [reply src/hash_table.py]: Make sense. I can do it like this, but isn't it verbose? @toet251 do you have any idea?
```
class ChainHashTable:
    class Node:
        def __init__(self, key: str, val: Union[str, float], next: ChainHashTable.Node = None):
            self.key = key
            self.val = val
            self.next = next

    def __init__(self, n: int):
        self.__bucket = [ChainHashTable.Node(None, None)] * n
```

File: tests/test_hash_table.py (hunk 6 above)
Comment (dvquy13) [reply tests/test_hash_table.py:6]: For Python built-in hash table `dict`, it will return KeyError when trying to remove from empty table. 
```
d_ = dict()
d_.pop("8") # KeyError
```
Should we mimic this behavior or just return None like what I have done?

Comment (trungducc) [reply tests/test_hash_table.py:6]: `Pop` will only return `KeyError` if the key doesn't present and default value not specified. Maybe you can try to add `default` parameter. I think it's up to you and @mindxhub/py team. I'm ok with current behavior.

**Comment by dvquy13** [comment]
@trungducc @toet251 Need your review man!
@trungducc after I push new commits, who will be the one to resolve the previous conversations and change requests? You or me or the reviewer?

File: src/hash_table.py (hunk 1 above)
Comment (toet251) [reply src/hash_table.py]: Idk. Maybe it is the only way for private class. Another accessing way is self.Node(), shorter anyway :))

**Comment by trungducc** [comment]
@dvquy13 For conversations, you should resolve them if you have permission. Change request will be approved by reviews after they review again.

File: src/hash_table.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+class Node:
+    def __init__(self, key: str, val: Union[str, float], next: Node = None):
+        self.key = key
+        self.val = val
+        self.next = next
+
+
+class ChainHashTable:
+    def __init__(self, n: int):
```
Comment (trungducc) [inline src/hash_table.py]: Parameter name should be `capacity` instead of `n`. We should also have default value for capacity in case user doesn't want to give a capacity.

File: src/hash_table.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+
+class Node:
+    def __init__(self, key: str, val: Union[str, float], next: Node = None):
+        self.key = key
+        self.val = val
+        self.next = next
+
+
+class ChainHashTable:
```
Comment (trungducc) [inline src/hash_table.py]: Did you forget making `Node` be a private class?

File: src/hash_table.py [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    def __init__(self, n: int):
+        self.__cap = n
+        self.__size = 0
+        self.__bucket = []
+        for _ in range(n):
+            self.__bucket.append(Node(None, None))
+
+    def cap(self) -> int:
```
Comment (trungducc) [inline src/hash_table.py]: I think we should use `capacity` instead of `cap`.

File: src/hash_table.py [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        return self.__cap
+
+    def size(self) -> int:
+        return self.__size
+
+    def insert(self, key: str, val: Union[str, float]) -> None:
```
Comment (trungducc) [inline src/hash_table.py]: Did you forget removing `-> None`?

File: src/hash_table.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+    def remove(self, key: str):
+        """
+        Remove a key-value pair if the key is found, else do nothing.
+        """
+        node: Node = self.__bucket[self.__hash(key)]
+        while node and node.next:
+            next_node = node.next
+            if next_node.key == key:
+                node.next = next_node.next
+                if self.size() > 0:
```
Comment (trungducc) [inline src/hash_table.py]: I think `size` will be always greater than 0 at this step.

File: src/hash_table.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+    def __hash(self, key: str) -> int:
+        """
+        Decide which bucket a key should belong to.
+        """
+        key: int = hash(key)
+        return key % len(self.__bucket)
+
+    def __table_doubling(self) -> None:
+        new_ht = ChainHashTable(self.cap() * 2)
```
Comment (trungducc) [inline src/hash_table.py]: I think creating an instance here is not normal. @toet251 How do you think?

File: tests/test_hash_table.py [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+        self.assertEqual(ht.search("11"), "c")
+        # Key not exist
+        self.assertEqual(ht.search("88"), None)
+
+    def test_remove(self):
+        ht = ChainHashTable(8)
+        # Remove from empty hash table should return None
+        self.assertEqual(ht.remove("8"), None)
+        # Remove from empty should leave size == 0 rather than -1
+        self.assertEqual(ht.size(), 0)
```
Comment (trungducc) [inline tests/test_hash_table.py:28]: IMO, this method needs to test for normal case. For example, when we remove keys which exist.

**Review by trungducc — CHANGES_REQUESTED** [review]

File: src/hash_table.py (hunk 12 above)
Comment (dvquy13) [reply src/hash_table.py]: I was waiting for QA's comment. Gonna update at the next round of commits.

File: src/hash_table.py (hunk 14 above)
Comment (dvquy13) [reply src/hash_table.py]: As Python implicitly `return None`, what I actually forgot was to add `return None` to method `remove`

Comment (dvquy13) [reply src/hash_table.py]: As per our recent discussion, I agree that we should remove `-> None`.

File: tests/test_hash_table.py (hunk 17 above)
Comment (dvquy13) [reply tests/test_hash_table.py:28]: I included tests for the normal cases in `test_all` method. I intended to only create new test method for edge cases, so I will change the name `test_all` to `test_normal`.

**Review by anhntt96 — CHANGES_REQUESTED** [review]
Why do we need to import **annotations** and **random** ?

File: src/hash_table.py (hunk 16 above)
Comment (dvquy13) [reply src/hash_table.py]: Agree, I will reimplement without initiating new instance

**Comment by dvquy13** [comment]
> Why do we need to import **annotations** and **random** ?

@anhntt96 Please review.

1. I have removed `import random`.
2. I need to `import annotations` to enable the calling of `Node` in this line:
```
def __init__(self, key: str, val: Union[str, float], next: ChainHashTable.Node = None):
```

**Comment by dvquy13** [comment]
@trungducc Please review

File: src/hash_table.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        node: self.Node = self.__bucket[self.__hash(key)]
+        while node.next:
+            node = node.next
+            if node.key == key:
+                node.val = val
+                return
+        node.next = self.Node(key, val)
+        self.__size += 1
+        if self.size() >= self.capacity() / 2:
```
Comment (trungducc) [inline src/hash_table.py]: I suggest to declaring a constant for `2` or `0.5`. We should not keep magic number in our code. In additional, if we want to change this number in the future, we don't have to look in to the code base if we use constants. Same with line 77.

**Review by trungducc — CHANGES_REQUESTED** [review]

File: src/hash_table.py (hunk 18 above)
Comment (dvquy13) [reply src/hash_table.py]: @trungducc got it. I just pushed a new commit to fix this.

**Comment by trungducc** [comment]
LGTM! Thanks @dvquy13 for your contribution!

**Review by trungducc — APPROVED** [review]

**Review by toet251 — APPROVED** [review]

**Comment by toet251** [comment]
@anhntt96 check it out bruh

**Comment by trungducc** [comment]
@anhntt96 With PRs which have more than 1 commit, we should use `Squash and merge` instead of `Create a merge commit`.

@dvquy13 Please help recreate this PR
